### PR TITLE
Implement token refresh interface and oauth token refreshing for GitHub

### PR DESCRIPTION
**Pull request by pjlast**
This is a quick refactor to have the `OAuthBearerToken` be a complete OAuth token, and use it instead of having another separate implementation of the OAuth token spec.

## Test plan

Existing tests should work, it's a refactor.

### CODE REVIEW

File: internal/extsvc/auth/oauth_bearer.go [hunk 1]
```diff
@@ -15,25 +15,28 @@ import (
 // OAuthBearerToken implements OAuth Bearer Token authentication for extsvc
 // clients.
 type OAuthBearerToken struct {
-	Token string
+	AccessToken  string    `json:"access_token"`
```
Comment (eseliger) [inline internal/extsvc/auth/oauth_bearer.go:18]: changing this from `Token` to `access_token` will break unmarshalling existing tokens afaict.

**Review by eseliger — CHANGES_REQUESTED** [review]

File: internal/database/oauth_token_helper.go [hunk 2]
```diff
@@ -15,76 +17,90 @@ import (
 // ExternalAccountTokenRefresher returns an oauthutil.TokenRefresher for the
 // given external account.
 func ExternalAccountTokenRefresher(db DB, externalAccountID int32, refreshToken string) oauthutil.TokenRefresher {
-	return func(ctx context.Context, doer httpcli.Doer, oauthCtx oauthutil.OAuthContext) (token string, err error) {
+	return func(ctx context.Context, doer httpcli.Doer, oauthCtx oauthutil.OAuthContext) (token *auth.OAuthBearerToken, err error) {
 		defer func() {
 			success := err == nil
 			gitlab.TokenRefreshCounter.WithLabelValues("external_account", strconv.FormatBool(success)).Inc()
 		}()
```
Comment (pjlast) [inline internal/database/oauth_token_helper.go:24]: @miveronese does it still make sense to have these `gitlab.TokenRefreshCounter` increments here?

Comment (miveronese) [reply internal/database/oauth_token_helper.go:24]: I am not sure about the original context around starting exporting to Prometheus these types of metrics, so I would prefer to double-check with the Repo Management team if this was supposed to be a temporary thing or not.

File: internal/database/oauth_token_helper.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ func ExternalAccountTokenRefresher(db DB, externalAccountID int32, refreshToken
...
 
-		refreshedToken, err := oauthutil.RetrieveToken(doer, oauthCtx, refreshToken, oauthutil.AuthStyleInParams)
+		refreshedToken, err := oauthutil.RetrieveToken(doer, oauthCtx, refreshToken)
```
Comment (miveronese) [inline internal/database/oauth_token_helper.go:26]: @unknwon , can you refresh my memory on why we needed this originally? I remember that it was necessary but the reason escaped my memory :D

File: internal/database/oauth_token_helper.go (hunk 2 above)
Comment (miveronese) [reply internal/database/oauth_token_helper.go:24]: My main concern here is if there's an alert that uses these metrics (or any Grafana dashboard that is fed with it). So removing them wold be risky.

Comment (pjlast) [reply internal/database/oauth_token_helper.go:24]: I think they need to be factored out. These increments are specific to GitLab and will trigger for every refresh at the moment. So we'll have to trigger them in the GitLab client and not generically for every Auth provider

Comment (miveronese) [reply internal/database/oauth_token_helper.go:24]: Agree -- these need to be moved in a way that they can be used by other auth providers not only GL

**Comment by miveronese** [comment]
And update: I'm not sure if this was triggered by the recent commits such as the removal of AuthStyles or by something else, but the refresh stopped working for GL only. I will do some more tests later against previous versions to double-check my hypothesis and see what's wrong.

**Comment by pjlast** [comment]
>Refresh_token can be an empty string when we get it from the config file (GL case). In that case, we should not return an error.

I don't think that's the correct conclusion. If the token is expired and no refresh token is present, that error should definitely be returned, or at least be treated as we would normally have treated an expired token

**Comment by pjlast** [comment]
@miveronese I just converted the GitLab case to use the refresh interface, it was simple enough

File: internal/extsvc/auth/oauth_bearer.go (hunk 1 above)
Comment (pjlast) [reply internal/extsvc/auth/oauth_bearer.go]: @eseliger I worked through the code and marhsalling of existing tokens won't be affected. All tokens from `user_external_accounts` are unmarshalled into `oauth2.Token`s (and are already stored as `access_token`. And unmarshalling from `external_services` does not unmarshal into an `OAuthBearerToken` either

Comment (eseliger) [reply internal/extsvc/auth/oauth_bearer.go]: Batch changes stores OAuthBearerTokens as well, I think these would be affected

Comment (pjlast) [reply internal/extsvc/auth/oauth_bearer.go]: Ah, gotcha, I missed that context. Thanks

**Pull request closed without merging**